File: src/pursuit.js
import { predictedTarget } from "./steering.js";
const unit = 22.4 / 576;

// active_player_ai: no-item, visible field-player branch (WIP sub_D742).
export function pursuit(s, i, random, distances, inMultiplier) {
  const p = s.players[i],
    b = s.ball;
  const j = s.controlled[1 - p.team],
    q = s.players[j];
  const attack = Math.floor(p.stats[0] / 2) > random;
  const opponent =
    q &&
    (b.owner === j ||
      inMultiplier ||
      (q.stun <= 0 && q.health > 0 && attack && distances[i] > distances[j]));
  const target = opponent ? q : b;
  const [tx, tz] = predictedTarget(
    target.x,
    target.z,
    opponent ? q.moveX || 0 : b.vx,
    opponent ? q.moveZ || 0 : b.vz,
    p.stats[7],
  );
  const dx = Math.round(tx / unit) - Math.round(p.x / unit);
  const dz = Math.round(tz / unit) - Math.round(p.z / unit);
  const x = Math.abs(dx) > Math.floor(Math.abs(dz) / 2) ? Math.sign(dx) : 0;
  const z = Math.abs(dz) > Math.floor(Math.abs(dx) / 2) ? Math.sign(dz) : 0;
  return { attack, x, z, tx, tz };
}

File: server/pursuit.go
package main

import "math"

// active_player_ai: selected field-player targets (WIP sub_D742 / sub_D97E).
func (s *State) pursuit(i, random int, distances *[18]int, inMultiplier bool) (interaction, float64, float64) {
	const unit = 22.4 / 576
	p, b := &s.Players[i], &s.Ball
	j := s.Controlled[1-p.Team]
	visible := s.worldInViewport(p.X, p.Z, 16)
	attack := visible && p.Stats[0]/2 > random
	x, z, vx, vz := b.X, b.Z, b.VX, b.VZ
	if visible && j >= 0 && j < len(s.Players) {
		q := &s.Players[j]
		if b.Owner == j || inMultiplier || (q.Stun <= 0 && q.Health > 0 && attack && distances[i] > distances[j]) {
			x, z, vx, vz = q.X, q.Z, q.moveX, q.moveZ
		}
	}
	if visible {
		for _, slot := range []int{0, 1, 6, 2, 3, 4, 5} {
			item := s.Pickups[slot]
			if item.Kind == 0 || item.Wait > 0 || !s.worldInViewport(item.X, item.Z, 0) || referenceDistance(item.X-p.X, item.Z-p.Z) == 0 {
				continue
			}
			if referenceDistance(item.X-b.X, item.Z-b.Z) <= distances[i] {
				x, z, vx, vz = item.X, item.Z, 0, 0
			}
			break
		}
	}
	tx, tz := predictedTarget(x, z, vx, vz, p.Stats[7])
	dx, dz := int(math.Round(tx/unit)-math.Round(p.X/unit)), int(math.Round(tz/unit)-math.Round(p.Z/unit))
	a := interaction{attack: attack}
	if absInt(dx) > absInt(dz)/2 {
		a.x = math.Copysign(1, float64(dx))
	}
	if absInt(dz) > absInt(dx)/2 {
		a.z = math.Copysign(1, float64(dz))
	}
	return a, tx, tz
}
